lib/marked.js

```diff
@@ -554,9 +554,72 @@ inline.normal = merge({}, inline);
 inline.pedantic = merge({}, inline.normal, {
   strong: /^__(?=\S)([\s\S]*?\S)__(?!_)|^\*\*(?=\S)([\s\S]*?\S)\*\*(?!\*)/,
   em: /^_(?=\S)([\s\S]*?\S)_(?!_)|^\*(?=\S)([\s\S]*?\S)\*(?!\*)/,
-  link: edit(/^!?\[(label)\]\(\s*<?([\s\S]*?)>?(?:\s+(['"][\s\S]*?['"]))?\s*\)/)
-    .replace('label', inline._label)
-    .getRegex(),
+  /* Original link re: /^!?\[(label)\]\(\s*<?([\s\S]*?)>?(?:\s+(['"][\s\S]*?['"]))?\s*\)/
+   * This captures the spec reasonably well but is vulnerable to REDOS.
+   * Instead we use a custom parser that follows the RegExp.exec semantics. */
+  link: {
+    exec: function (s) {
+      // [TEXT](DESTINATION)
+      var generalLinkRe = edit(/^!?\[(label)\]\((.*?)\)/)
+        .replace('label', inline._label)
+        .getRegex();
+
+      // destination: DESTINATION from generalLinkRe
+      // returns [destination, title]: no angle-brackets on destination, no quotes on title
+      function splitIntoDestinationAndTitle (destination) {
+        function unwrapAngleBrackets (str) {
+          if (str.match(/^<.*>$/)) {
+            str = str.slice(1, -1);
+          }
+          return str;
+        }
+
+        // Valid DESTINATIONs, in decreasing specificity.
+        var destinationAndTitleRe = /^([^'"(]*[^\s])\s+(['"(].*['")])/;
+        var destinationRe = /^(<?[\s\S]*>?)/;
+        var parsingRegexes = [destinationAndTitleRe, destinationRe];
+
+        var match = false;
+        var dest = undefined;
+        var title = undefined;
+        for (var i = 0; i < parsingRegexes.length; i++) {
+          match = parsingRegexes[i].exec(destination);
+          if (match) {
+            dest = match[1];
+            title = match[2];
+            break;
+          }
+        }
+
+        if (match) {
+          // title is optional.
+          if (typeof title === 'undefined') {
+            title = '';
+          }
+
+          // Format dest.
+          dest = dest.trim();
+          dest = unwrapAngleBrackets(dest);
+
+          return [dest, title];
+        }
+        return null;
+      }
+
+      var fullMatch = generalLinkRe.exec(s);
+      if (fullMatch) {
```

Comment:
Can you flip this so that the if statement is smaller:

```
if (!fullMatch) {
  return null;
}
// ... split and such here
return [fullMatch[0], text, destinationAndTitle[0], destinationAndTitle[1]];
```